File: typescript/packages/jumpstarter-client/src/UdpPortforwardAdapter.ts
/**
 * Creates a local UDP socket that forwards datagrams over a gRPC
 * `RouterService.Stream` to a remote driver method.
 *
 * Each incoming datagram maps to one StreamChannel message, preserving
 * message boundaries.
 */

import * as dgram from "dgram";
import type { ExporterSession } from "./ExporterSession";
import { openDriverStream } from "./streamProto";
import type { StreamChannel } from "./StreamChannel";

export class UdpPortforwardAdapter {
  private readonly socket: dgram.Socket;
  private readonly _address: string;
  private readonly _port: number;
  private channel: StreamChannel | null = null;
  private readLoop: Promise<void> | null = null;

  private constructor(socket: dgram.Socket, address: string, port: number) {
    this.socket = socket;
    this._address = address;
    this._port = port;
  }

  /** The local address the UDP socket is bound to. */
  get address(): string {
    return this._address;
  }

  /** The local port the UDP socket is bound to. */
  get port(): number {
    return this._port;
  }

  /**
   * Open a UDP port-forward adapter that listens on an ephemeral local port.
   *
   * Opens a single `RouterService.Stream` and forwards datagrams bidirectionally.
   * Each UDP datagram sent to the local port becomes one gRPC stream message,
   * and each incoming stream message is sent back as a datagram to the last
   * known remote address.
   *
   * @param session - The exporter session providing address and credentials
   * @param driverUuid - UUID of the target driver instance
   * @param method - The driver method to invoke (e.g. "connect")
   * @returns A started UdpPortforwardAdapter
   */
  static async open(
    session: ExporterSession,
    driverUuid: string,
    method: string,
  ): Promise<UdpPortforwardAdapter> {
    const socket = dgram.createSocket("udp4");

    const host = "127.0.0.1";
    await new Promise<void>((resolve, reject) => {
      socket.bind(0, host, () => resolve());
      socket.once("error", reject);
    });

    const addr = socket.address();
    const adapter = new UdpPortforwardAdapter(socket, addr.address, addr.port);

    const { channel } = openDriverStream(session, driverUuid, method);
    adapter.channel = channel;

    // Track the last peer so we can send responses back
    let lastRemotePort = 0;
    let lastRemoteAddress = "";

    // Forward incoming UDP datagrams → gRPC stream
    socket.on("message", (msg: Buffer, rinfo: dgram.RemoteInfo) => {
      lastRemotePort = rinfo.port;
      lastRemoteAddress = rinfo.address;
      channel.write(msg).catch(() => {
        // stream closed, ignore
      });
    });

    // Forward gRPC stream → UDP datagrams back to last known peer
    adapter.readLoop = (async () => {
      try {
        for await (const chunk of channel) {
          if (lastRemotePort > 0) {
            socket.send(chunk, lastRemotePort, lastRemoteAddress);
          }
        }
      } catch {
        // stream ended or errored
      }
    })();

    return adapter;
  }

  /** Close the UDP socket and the gRPC stream. */
  close(): void {
    this.channel?.close();
    this.channel = null;
    try {
      this.socket.close();
    } catch {
      // already closed
    }
  }
}

File: java/jumpstarter-client/src/main/java/dev/jumpstarter/client/UdpPortforwardAdapter.java
package dev.jumpstarter.client;

import jumpstarter.v1.RouterServiceGrpc;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.io.InputStream;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetSocketAddress;
import java.net.SocketAddress;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Creates a local UDP listener that forwards datagrams over a
 * {@link StreamChannel} to a remote driver's stream method.
 *
 * <p>Each unique remote address (client) gets its own {@link StreamChannel}.
 * Each datagram maps to one StreamChannel message.
 *
 * <p>Usage:
 * <pre>{@code
 * try (UdpPortforwardAdapter adapter = UdpPortforwardAdapter.open(session, driverUuid, "connect")) {
 *     InetSocketAddress addr = adapter.getLocalAddress();
 *     // Send/receive UDP datagrams to addr
 * }
 * }</pre>
 */
public final class UdpPortforwardAdapter implements AutoCloseable {

    private static final Logger logger = Logger.getLogger(UdpPortforwardAdapter.class.getName());
    private static final int BUFFER_SIZE = 65535;

    private final DatagramSocket socket;
    private final Map<SocketAddress, StreamChannel> channels = new ConcurrentHashMap<>();
    private volatile boolean closed = false;

    private UdpPortforwardAdapter(@NotNull DatagramSocket socket) {
        this.socket = socket;
    }

    /**
     * Open a UDP port-forward adapter that listens on an ephemeral local port.
     *
     * <p>Incoming datagrams are forwarded to the specified driver stream method.
     * Responses from the stream are sent back as datagrams to the original sender.
     *
     * @param session    the exporter session providing the gRPC channel
     * @param driverUuid the target driver instance UUID
     * @param method     the stream method name (e.g. "connect")
     * @return a running adapter with an active local UDP listener
     */
    @NotNull
    public static UdpPortforwardAdapter open(
            @NotNull ExporterSession session,
            @NotNull String driverUuid,
            @NotNull String method) {
        try {
            DatagramSocket ds = new DatagramSocket(new InetSocketAddress("127.0.0.1", 0));
            UdpPortforwardAdapter adapter = new UdpPortforwardAdapter(ds);

            Thread recvThread = new Thread(() -> adapter.receiveLoop(session, driverUuid, method),
                    "jumpstarter-udpfwd-recv-" + ds.getLocalPort());
            recvThread.setDaemon(true);
            recvThread.start();

            logger.info("UDP port-forward listening on " + ds.getLocalSocketAddress()
                    + " → driver " + driverUuid + "." + method);
            return adapter;
        } catch (IOException e) {
            throw new RuntimeException("Failed to create UDP port-forward listener", e);
        }
    }

    /**
     * Get the local address and port the listener is bound to.
     *
     * @return the local socket address
     */
    @NotNull
    public InetSocketAddress getLocalAddress() {
        return (InetSocketAddress) socket.getLocalSocketAddress();
    }

    @Override
    public void close() {
        if (closed) return;
        closed = true;
        socket.close();
        for (StreamChannel ch : channels.values()) {
            try { ch.close(); } catch (Exception ignored) {}
        }
        channels.clear();
    }

    private void receiveLoop(ExporterSession session, String driverUuid, String method) {
        RouterServiceGrpc.RouterServiceStub routerStub =
                RouterServiceGrpc.newStub(session.getChannel());
        byte[] buf = new byte[BUFFER_SIZE];

        while (!closed) {
            try {
                DatagramPacket packet = new DatagramPacket(buf, buf.length);
                socket.receive(packet);

                SocketAddress sender = packet.getSocketAddress();
                StreamChannel channel = channels.computeIfAbsent(sender, addr -> {
                    StreamChannel ch = StreamChannel.open(routerStub, driverUuid, method);
                    startReturnThread(addr, ch);
                    return ch;
                });

                byte[] data = new byte[packet.getLength()];
                System.arraycopy(packet.getData(), packet.getOffset(), data, 0, packet.getLength());
                channel.outputStream().write(data);
            } catch (IOException e) {
                if (!closed) {
                    logger.log(Level.WARNING, "UDP port-forward receive error", e);
                }
            }
        }
    }

    private void startReturnThread(SocketAddress sender, StreamChannel channel) {
        Thread t = new Thread(() -> {
            try {
                InputStream in = channel.inputStream();
                byte[] buf = new byte[BUFFER_SIZE];
                int n;
                while ((n = in.read(buf)) != -1) {
                    DatagramPacket reply = new DatagramPacket(buf, 0, n, sender);
                    socket.send(reply);
                }
            } catch (IOException e) {
                if (!closed) {
                    logger.log(Level.FINE, "UDP return stream ended", e);
                }
            } finally {
                channels.remove(sender);
                try { channel.close(); } catch (Exception ignored) {}
            }
        }, "jumpstarter-udpfwd-ret-" + sender);
        t.setDaemon(true);
        t.start();
    }
}
